TrackletAlgorithm/TrackBuilder.h

```diff
@@ -292,46 +301,44 @@ void TrackBuilder(
     }
 
     // Only tracks with at least two matches are valid.
-    track.setTrackValid(nMatches >= kMinNMatches);
+    track.setTrackValid(!done && (nMatches >= kMinNMatches));
 
     // Output the track word and eight stub words associated with the TrackFit
     // object that was constructed.
-    if (track.getTrackValid()) {
-      trackWord[nTracks] = track.getTrackWord();
-      barrel_stub_words: for (short j = 0 ; NBarrelStubs > 0 && j < NBarrelStubs; j++) { // Note: need to have NBarrelStubs > 0 to prevent compilation error due to -Werror=type-limits flag in CMSSW
-        switch (j) {
-          case 0:
-            barrelStubWords[j][nTracks] = track.template getStubValid<0>() ? track.template getBarrelStubWord<0>() : typename TrackFit<NBarrelStubs, NDiskStubs>::BarrelStubWord(0);
-            break;
-          case 1:
-            barrelStubWords[j][nTracks] = track.template getStubValid<1>() ? track.template getBarrelStubWord<1>() : typename TrackFit<NBarrelStubs, NDiskStubs>::BarrelStubWord(0);
-            break;
-          case 2:
-            barrelStubWords[j][nTracks] = track.template getStubValid<2>() ? track.template getBarrelStubWord<2>() : typename TrackFit<NBarrelStubs, NDiskStubs>::BarrelStubWord(0);
-            break;
-          case 3:
-            barrelStubWords[j][nTracks] = track.template getStubValid<3>() ? track.template getBarrelStubWord<3>() : typename TrackFit<NBarrelStubs, NDiskStubs>::BarrelStubWord(0);
-            break;
-        }
+    trackWord[nTracks] = track.getTrackWord();
+    barrel_stub_words: for (short j = 0 ; NBarrelStubs > 0 && j < NBarrelStubs; j++) { // Note: need to have NBarrelStubs > 0 to prevent compilation error due to -Werror=type-limits flag in CMSSW
+      switch (j) {
+        case 0:
+          barrelStubWords[j][nTracks] = track.template getBarrelStubWord<0>();
+          break;
+        case 1:
+          barrelStubWords[j][nTracks] = track.template getBarrelStubWord<1>();
+          break;
+        case 2:
+          barrelStubWords[j][nTracks] = track.template getBarrelStubWord<2>();
+          break;
+        case 3:
+          barrelStubWords[j][nTracks] = track.template getBarrelStubWord<3>();
+          break;
```

Comment:
Just curious, and not something that needs addressing in this PR necessarily but Is there a reason why `getBarrelStubWord` (and others like it defined in TrackFitMemory) have  `Hit` as a template instead of function parameter? is it  possible to have this loop(and others like it) just do something like: 
`barrelStubWords[j][nTracks] = track.template getBarrelStubWord(j);` without the switch?

Reply:
I like to be as explicit as possible about what is known at compile time and make those variables template parameters. Here, since the loop is unrolled, each of the "hit" numbers is known at compile time.

It's entirely possible that HLS is smart enough to infer that these numbers are known at compile time, so that if they were passed as regular function arguments, they would be treated the same as they are now. But that would require reworking quite a bit of TrackFitMemory. So it might result in cleaner code, but I'm not sure it's worth trying right now.

Reply:
Ok, that makes sense, let's not worry about this now then.